pkg/kubelet/stats/stats_provider.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stats
+
+import (
+	"fmt"
+
+	cadvisorapiv1 "github.com/google/cadvisor/info/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	internalapi "k8s.io/kubernetes/pkg/kubelet/apis/cri"
+	statsapi "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+	"k8s.io/kubernetes/pkg/kubelet/cadvisor"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	kubepod "k8s.io/kubernetes/pkg/kubelet/pod"
+	"k8s.io/kubernetes/pkg/kubelet/server/stats"
+)
+
+// NewCRIStatsProvider returns a StatsProvider that provides the node stats
+// from cAdvisor and the container stats from CRI.
+func NewCRIStatsProvider(
+	cadvisor cadvisor.Interface,
+	resourceAnalyzer stats.ResourceAnalyzer,
+	podManager kubepod.Manager,
+	runtimeCache kubecontainer.RuntimeCache,
+	runtimeService internalapi.RuntimeService,
+	imageService internalapi.ImageManagerService,
+) *StatsProvider {
+	return newStatsProvider(cadvisor, podManager, runtimeCache, newCRIStatsProvider(cadvisor, resourceAnalyzer, runtimeService, imageService))
+}
+
+// NewCadvisorStatsProvider returns a containerStatsProvider that provides both
+// the node and the container stats from cAdvisor.
+func NewCadvisorStatsProvider(
+	cadvisor cadvisor.Interface,
+	resourceAnalyzer stats.ResourceAnalyzer,
+	podManager kubepod.Manager,
+	runtimeCache kubecontainer.RuntimeCache,
+	imageService kubecontainer.ImageService,
+) *StatsProvider {
+	return newStatsProvider(cadvisor, podManager, runtimeCache, newCadvisorStatsProvider(cadvisor, resourceAnalyzer, imageService))
+}
+
+// newStatsProvider returns a new StatsProvider that provides node stats from
+// cAdvisor and the container stats using the containerStatsProvider.
+func newStatsProvider(
+	cadvisor cadvisor.Interface,
+	podManager kubepod.Manager,
+	runtimeCache kubecontainer.RuntimeCache,
+	containerStatsProvider containerStatsProvider,
+) *StatsProvider {
+	return &StatsProvider{
+		cadvisor:               cadvisor,
+		podManager:             podManager,
+		runtimeCache:           runtimeCache,
+		containerStatsProvider: containerStatsProvider,
+	}
+}
+
+// StatsProvider provides the stats of the node and the pod-managed containers.
+type StatsProvider struct {
```

Comment:
This type is called `StatsProvider` instead of `Provider` because it is embedded in `Kubelet`. If we call it `Provider` we will use it in kubelet as `kubelet.Provider`, which is confusing.

Reply:
It does not fully implement the interface - functions like `GetPodByName()`, `GetNode()` in the interface are implemented by the `Kubelet`.